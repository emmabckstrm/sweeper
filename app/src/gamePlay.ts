import { getAdjacentSquares } from "./setup";
import type { GameBoard, SquareState, SquareStatus } from "./types";

export const updateBoardWithSquare = (
  board: GameBoard,
  row: number,
  col: number,
  newValue: Partial<SquareStatus>
) => {
  const newBoard = board.slice();
  newBoard[row][col] = { ...newBoard[row][col], ...newValue };
  return newBoard;
};

export const getBoardWithOpenedSquares = (
  row: number,
  col: number,
  _gameBoard: GameBoard
): GameBoard => {
  let gameBoard = _gameBoard.slice();
  const square = gameBoard[row][col];

  console.log("square", square);
  if (!isOpen(square) && !square.isBomb && !isFlagged(square)) {
    if (square.adjacentBombs > 0) {
      gameBoard = updateBoardWithSquare(gameBoard, row, col, {
        status: "open",
      });
    } else {
      gameBoard = updateBoardWithSquare(gameBoard, row, col, {
        status: "open",
      });
      const adjacents = getAdjacentSquares(row, col, gameBoard);
      for (let i = 0; i < adjacents.length; i++) {
        gameBoard = getBoardWithOpenedSquares(
          adjacents[i][0],
          adjacents[i][1],
          gameBoard
        );
      }
    }
  }
  return gameBoard;
};

export const getNumberOfOpenSquares = (board: GameBoard) => {
  let numberOfOpenedSquares = 0;
  board.forEach((row) => {
    row.forEach((square) => {
      if (isOpen(square)) {
        numberOfOpenedSquares += 1;
      }
    });
  });
  return numberOfOpenedSquares;
};

export const setAllSquaresToOpen = (_board: GameBoard) => {
  const board = _board.slice();
  board.forEach((row, r) => {
    row.forEach((square, c) => {
      if (!isOpen(square)) {
        square.status = "open";
      }
    });
  });

  return board;
};

export const isOpen = (square: SquareStatus) => {
  return square.status === "open";
};
export const isFlagged = (square: SquareStatus) => {
  return square.status === "flag";
};
