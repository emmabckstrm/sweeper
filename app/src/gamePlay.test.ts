import {
  initGameBoard,
  positionExistsInArray,
  generateBombPositions,
  getAdjacentSquares,
} from "./setup";
import { getBoardWithOpenedSquares } from "./gamePlay";

describe(getBoardWithOpenedSquares, () => {
  const gameBoard = initGameBoard(3, 3);
  gameBoard[0][0] = { ...gameBoard[0][0], isBomb: true };
  const adjacents = getAdjacentSquares(0, 0, gameBoard);
  adjacents.forEach((adjacent) => {
    gameBoard[adjacent[0]][adjacent[1]] = {
      ...gameBoard[adjacent[0]][adjacent[1]],
      adjacentBombs: 1,
    };
  });

  it("only opens one if it has adjacentBombs number", () => {
    expect(getBoardWithOpenedSquares(0, 1, gameBoard)).toMatchSnapshot();
  });

  it("opens all adjacent empty and adjacent with adjacentBombs number", () => {
    expect(getBoardWithOpenedSquares(2, 2, gameBoard)).toMatchSnapshot();
  });

  it("doesn't open flagged squares", () => {
    gameBoard[0][2].isFlagged = true;
    expect(getBoardWithOpenedSquares(2, 2, gameBoard)).toMatchSnapshot();
  });
});
