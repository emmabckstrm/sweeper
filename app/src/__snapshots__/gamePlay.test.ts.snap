// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getBoardWithOpenedSquares doesn't open flagged squares 1`] = `
[
  [
    {
      "adjacentBombs": 0,
      "isBomb": true,
      "isFlagged": false,
      "isOpen": false,
    },
    {
      "adjacentBombs": 1,
      "isBomb": false,
      "isFlagged": false,
      "isOpen": true,
    },
    {
      "adjacentBombs": 0,
      "isBomb": false,
      "isFlagged": true,
      "isOpen": false,
    },
  ],
  [
    {
      "adjacentBombs": 1,
      "isBomb": false,
      "isFlagged": false,
      "isOpen": true,
    },
    {
      "adjacentBombs": 1,
      "isBomb": false,
      "isFlagged": false,
      "isOpen": true,
    },
    {
      "adjacentBombs": 0,
      "isBomb": false,
      "isFlagged": false,
      "isOpen": true,
    },
  ],
  [
    {
      "adjacentBombs": 0,
      "isBomb": false,
      "isFlagged": false,
      "isOpen": true,
    },
    {
      "adjacentBombs": 0,
      "isBomb": false,
      "isFlagged": false,
      "isOpen": true,
    },
    {
      "adjacentBombs": 0,
      "isBomb": false,
      "isFlagged": false,
      "isOpen": true,
    },
  ],
]
`;

exports[`getBoardWithOpenedSquares only opens one if it has adjacentBombs number 1`] = `
[
  [
    {
      "adjacentBombs": 0,
      "isBomb": true,
      "isFlagged": false,
      "isOpen": false,
    },
    {
      "adjacentBombs": 1,
      "isBomb": false,
      "isFlagged": false,
      "isOpen": true,
    },
    {
      "adjacentBombs": 0,
      "isBomb": false,
      "isFlagged": false,
      "isOpen": false,
    },
  ],
  [
    {
      "adjacentBombs": 1,
      "isBomb": false,
      "isFlagged": false,
      "isOpen": false,
    },
    {
      "adjacentBombs": 1,
      "isBomb": false,
      "isFlagged": false,
      "isOpen": false,
    },
    {
      "adjacentBombs": 0,
      "isBomb": false,
      "isFlagged": false,
      "isOpen": false,
    },
  ],
  [
    {
      "adjacentBombs": 0,
      "isBomb": false,
      "isFlagged": false,
      "isOpen": false,
    },
    {
      "adjacentBombs": 0,
      "isBomb": false,
      "isFlagged": false,
      "isOpen": false,
    },
    {
      "adjacentBombs": 0,
      "isBomb": false,
      "isFlagged": false,
      "isOpen": false,
    },
  ],
]
`;

exports[`getBoardWithOpenedSquares opens all adjacent empty and adjacent with adjacentBombs number 1`] = `
[
  [
    {
      "adjacentBombs": 0,
      "isBomb": true,
      "isFlagged": false,
      "isOpen": false,
    },
    {
      "adjacentBombs": 1,
      "isBomb": false,
      "isFlagged": false,
      "isOpen": true,
    },
    {
      "adjacentBombs": 0,
      "isBomb": false,
      "isFlagged": false,
      "isOpen": true,
    },
  ],
  [
    {
      "adjacentBombs": 1,
      "isBomb": false,
      "isFlagged": false,
      "isOpen": true,
    },
    {
      "adjacentBombs": 1,
      "isBomb": false,
      "isFlagged": false,
      "isOpen": true,
    },
    {
      "adjacentBombs": 0,
      "isBomb": false,
      "isFlagged": false,
      "isOpen": true,
    },
  ],
  [
    {
      "adjacentBombs": 0,
      "isBomb": false,
      "isFlagged": false,
      "isOpen": true,
    },
    {
      "adjacentBombs": 0,
      "isBomb": false,
      "isFlagged": false,
      "isOpen": true,
    },
    {
      "adjacentBombs": 0,
      "isBomb": false,
      "isFlagged": false,
      "isOpen": true,
    },
  ],
]
`;
